segfault in thread/test_async_timer_mt.multithread/0

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/655/label=master/console
```
12:47:43 [----------] 1 test from thread/test_async_timer_mt
12:47:43 [ RUN      ] thread/test_async_timer_mt.multithread/0
12:47:43 [jenkins03:7099 :10] Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
12:47:43 make: *** [test] Segmentation fault
12:47:43 make: Leaving directory `/scrap/jenkins/jobs/gh-ucx-pr__NEW/workspace-5/label/master/build-test/test/gtest'
```

--- a/src/ucs/async/async.c
+++ b/src/ucs/async/async.c
@@ -255,14 +255,20 @@ ucs_status_t ucs_async_dispatch_handlers(int *events, size_t count)
 ucs_status_t ucs_async_dispatch_timerq(ucs_timer_queue_t *timerq,
                                        ucs_time_t current_time)
 {
-    size_t num_timers = 0;
+    size_t max_timers, num_timers = 0;
     int *expired_timers;
     ucs_timer_t *timer;
 
-    expired_timers = ucs_alloca(ucs_timerq_size(timerq) * sizeof(*expired_timers));
-    ucs_timerq_for_each_expired(timer, timerq, current_time) {
+    max_timers     = ucs_max(1, ucs_timerq_size(timerq));
+    expired_timers = ucs_alloca(max_timers * sizeof(*expired_timers));
+
+    ucs_timerq_for_each_expired(timer, timerq, current_time, {
         expired_timers[num_timers++] = timer->id;
-    }
+        if (num_timers >= max_timers) {
+            break; /* Keep timers which we don't have room for in the queue */
+        }
+    })
+
     return ucs_async_dispatch_handlers(expired_timers, num_timers);
 }
 

--- a/src/ucs/async/thread.c
+++ b/src/ucs/async/thread.c
@@ -51,6 +51,9 @@ static void ucs_async_thread_hold(ucs_async_thread_t *thread)
 static void ucs_async_thread_put(ucs_async_thread_t *thread)
 {
     if (ucs_atomic_fadd32(&thread->refcnt, -1) == 1) {
+        close(thread->epfd);
+        ucs_async_pipe_destroy(&thread->wakeup);
+        ucs_timerq_cleanup(&thread->timerq);
         ucs_free(thread);
     }
 }
@@ -120,9 +123,6 @@ static void *ucs_async_thread_func(void *arg)
         }
     }
 
-    close(thread->epfd);
-    ucs_async_pipe_destroy(&thread->wakeup);
-    ucs_timerq_cleanup(&thread->timerq);
     ucs_async_thread_put(thread);
     return NULL;
 }

--- a/src/ucs/time/timerq.h
+++ b/src/ucs/time/timerq.h
@@ -98,21 +98,21 @@ static inline int ucs_timerq_is_empty(ucs_timer_queue_t *timerq) {
  * @note Timers which expired between calls to this function will also be dispatched.
  * @note There is no guarantee on the order of dispatching.
  */
-#define ucs_timerq_for_each_expired(_timer, _timerq, _current_time) \
-    for ( \
-         /* Initialization */ \
-         pthread_spin_lock(&(_timerq)->lock), /* Grab lock */ \
-         _timer = (_timerq)->timers; /* Set iterator */ \
-         \
-         /* Condition */ \
-         (_timer != (_timerq)->timers + (_timerq)->num_timers) || /* Reached the end */ \
-         pthread_spin_unlock(&(_timerq)->lock); /* Release lock - should return 0 */ \
-         \
-         /* Step: advance iterator */ \
-         ++_timer) \
-        \
-        if ((_current_time >= (_timer)->expiration) /* Check if this timer has expirted */ && \
-            ((_timer)->expiration = _current_time + (_timer)->interval) /* Update expiration time */ \
-           )
+#define ucs_timerq_for_each_expired(_timer, _timerq, _current_time, _code) \
+    { \
+        ucs_time_t __current_time = _current_time; \
+        pthread_spin_lock(&(_timerq)->lock); /* Grab lock */ \
+        for (_timer = (_timerq)->timers; \
+             _timer != (_timerq)->timers + (_timerq)->num_timers; \
+             ++_timer) \
+        { \
+            if (__current_time >= (_timer)->expiration) { \
+                /* Update expiration time */ \
+                (_timer)->expiration = __current_time + (_timer)->interval; \
+                _code; \
+            } \
+        } \
+        pthread_spin_unlock(&(_timerq)->lock); /* Release lock  */ \
+    }
 
 #endif

--- a/test/gtest/ucs/test_time.cc
+++ b/test/gtest/ucs/test_time.cc
@@ -92,10 +92,10 @@ UCS_TEST_F(test_time, timerq) {
         counter2 = 0;
         for (unsigned count = 0; count < test_time; ++count) {
             ++current_time;
-            ucs_timerq_for_each_expired(timer, &timerq, current_time) {
+            ucs_timerq_for_each_expired(timer, &timerq, current_time, {
                 if (timer->id == TIMER_ID_1) ++counter1;
                 if (timer->id == TIMER_ID_2) ++counter2;
-            }
+            })
         }
         EXPECT_NEAR(test_time / interval1, counter1, 1);
         EXPECT_NEAR(test_time / interval2, counter2, 1);
@@ -109,10 +109,10 @@ UCS_TEST_F(test_time, timerq) {
         ASSERT_UCS_OK(status);
         for (unsigned count = 0; count < test_time; ++count) {
             ++current_time;
-            ucs_timerq_for_each_expired(timer, &timerq, current_time) {
+            ucs_timerq_for_each_expired(timer, &timerq, current_time, {
                 if (timer->id == TIMER_ID_1) ++counter1;
                 if (timer->id == TIMER_ID_2) ++counter2;
-            }
+            })
         }
         EXPECT_EQ(0u, counter1);
         EXPECT_NEAR(test_time / interval2, counter2, 1);
@@ -126,10 +126,10 @@ UCS_TEST_F(test_time, timerq) {
         counter2 = 0;
         for (unsigned count = 0; count < test_time; ++count) {
             ++current_time;
-            ucs_timerq_for_each_expired(timer, &timerq, current_time) {
+            ucs_timerq_for_each_expired(timer, &timerq, current_time, {
                 if (timer->id == TIMER_ID_1) ++counter1;
                 if (timer->id == TIMER_ID_2) ++counter2;
-            }
+            })
         }
         EXPECT_NEAR(test_time / interval1, counter1, 1);
         EXPECT_NEAR(test_time / interval2, counter2, 1);

